src/query/functions/src/scalars/decimal.rs

```diff
@@ -927,13 +1219,102 @@ fn decimal_to_decimal(
     }
 }
 
+macro_rules! single_decimal_to_decimal {
+    ($from: expr, $from_size: expr, $dest_size: expr, $from_type_name: ty, $dest_type_name: ty) => {{
+        // faster path
+        if $from_size.scale == $dest_size.scale && $from_size.precision <= $dest_size.precision {
+            // 128 -> 128 or 128 -> 256 or 256 -> 256
+            Some($from * <$dest_type_name>::one())
+        } else if $from_size.scale > $dest_size.scale {
+            let factor = <$dest_type_name>::e(($from_size.scale - $dest_size.scale) as u32);
+            let x = $from * <$dest_type_name>::one();
+            x.checked_div(factor)
+        } else {
+            let factor = <$dest_type_name>::e(($dest_size.scale - $from_size.scale) as u32);
+            let max = <$dest_type_name>::max_for_precision($dest_size.precision);
+            let min = <$dest_type_name>::min_for_precision($dest_size.precision);
+            let x = $from * <$dest_type_name>::one();
+            match x.checked_mul(factor) {
+                Some(x) if x <= max && x >= min => Some(x),
+                _ => None,
+            }
+        }
+    }};
+}
+
+fn single_decimal_256_to_128(
```

Comment:
what's the difference between:
https://github.com/andylokandy/databend/blob/811ff87f748ae5ba325afa6c930605cc40fa1ab8/src/query/functions/src/scalars/decimal.rs#L884

Reply:
`single_decimal_256_to_128` only operates on one decimal.

Reply:
so, can the new one be a wrapper of the old one?

Reply:
I‘m afraid not. The new method is a scalar method, and the old one is for vector. If we reuse the codes, we will introduce instructions and branches into the vectorized loop.